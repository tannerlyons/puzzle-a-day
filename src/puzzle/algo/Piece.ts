type Permutation = boolean[][];

const parseLayout = (layout: string[]): Permutation[] => {
    function parse(layout: string[]): Permutation {
        return layout.map((line) => line.split("").map((char) => char === "X"));
    }

    function rotate(
        parsedLayout: Permutation,
        rotateCount: 1 | 2 | 3
    ): Permutation {
        let prev = parsedLayout;
        let rotated = parsedLayout;

        for (let numRotations = 0; numRotations < rotateCount; numRotations++) {
            const oldDimensions = {
                x: prev[0].length,
                y: prev.length,
            };

            rotated = new Array(oldDimensions.x)
                .fill(1)
                .map((i) => new Array(oldDimensions.y).fill(1));

            for (let oldRowIdx = 0; oldRowIdx < oldDimensions.y; oldRowIdx++) {
                for (
                    let oldColIdx = 0;
                    oldColIdx < oldDimensions.x;
                    oldColIdx++
                ) {
                    const newYIdx = oldColIdx;
                    const newXIdx = oldDimensions.y - 1 - oldRowIdx;
                    rotated[newYIdx][newXIdx] = prev[oldRowIdx][oldColIdx];
                }
            }

            prev = rotated;
        }
        return rotated;
    }

    function validate(layout: string[]): void {
        const firstLen = layout[0].length;
        for (let i = 1; i < layout.length; i++) {
            const line = layout[i];
            if (firstLen !== line.length) {
                throw new Error("Invalid layout: " + layout.join(","));
            }
        }
    }

    function removeDuplicates(perms: Permutation[]): Permutation[] {
        function hash(perm: Permutation) {
            return perm
                .map((line) => line.map((c) => (c ? "X" : " ")).join(","))
                .join("n");
        }
        const hashes = perms.map(hash);
        const seen = new Set<string>();
        const ret: Permutation[] = [];
        perms.forEach((perm, idx) => {
            if (!seen.has(hashes[idx])) {
                seen.add(hashes[idx]);
                ret.push(perm);
            }
        });
        return ret;
    }

    function flip(permutation: Permutation) {
        return permutation.map((line) => line.slice().reverse());
    }

    validate(layout);
    const parsed = parse(layout);
    const permutations: Permutation[] = [
        parsed,
        rotate(parsed, 1),
        rotate(parsed, 2),
        rotate(parsed, 3),
        flip(parsed),
        flip(rotate(parsed, 1)),
        flip(rotate(parsed, 2)),
        flip(rotate(parsed, 3)),
    ];

    return removeDuplicates(permutations);
};

export class Piece {
    private permutations: Permutation[];

    constructor(public readonly name: string, layout: string[]) {
        this.permutations = parseLayout(layout);
    }

    public debug(): string {
        return (
            this.name +
            ":\n" +
            "-------------\n" +
            this.permutations
                .map(
                    (perm, idx) =>
                        `${1 + idx}. ` +
                        perm
                            .map(
                                (line, lineIdx) =>
                                    `${lineIdx === 0 ? "" : "   "}|` +
                                    line
                                        .map((ch) => (ch ? "X" : " "))
                                        .join("") +
                                    "|"
                            )
                            .join("\n")
                )
                .join("\n\n")
        );
    }
}

export const allPieces = [
    /**
     * bucket
     *
     * X X
     * XXX
     *
     */
    new Piece("bucket", ["X X", "XXX"]),

    /**
     * brick
     *
     * XX
     * XX
     * XX
     *
     */
    new Piece("brick", ["XX", "XX", "XX"]),

    /**
     * snek
     *
     * XX
     *  X
     *  XX
     *
     */
    new Piece("snek", ["XX ", " X ", " XX"]),

    /**
     * fork
     *
     * X
     * X
     * XX
     *  X
     *
     */
    new Piece("fork", ["X ", "X ", "XX", " X"]),

    /**
     * elle
     *
     *   X
     *   X
     * XXX
     *
     */
    new Piece("elle", ["  X", "  X", "XXX"]),

    /**
     * cactus
     *
     * X
     * XX
     * X
     * X
     *
     */
    new Piece("cactus", ["X ", "XX", "X ", "X "]),

    /**
     * utah
     *
     * X
     * XX
     * XX
     *
     */
    new Piece("utah", ["X ", "XX", "XX"]),

    /**
     * hoe
     *
     * XX
     * X
     * X
     * X
     *
     */
    new Piece("hoe", ["XX", "X ", "X ", "X "]),
];
