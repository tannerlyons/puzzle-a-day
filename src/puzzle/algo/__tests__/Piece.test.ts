import { allPieces } from "../Piece";

describe("Piece", () => {
    it("creates all combinations", () => {
        for (const piece of allPieces) {
            expect(piece.debug()).toMatchSnapshot(piece.name);
        }
    });
});
