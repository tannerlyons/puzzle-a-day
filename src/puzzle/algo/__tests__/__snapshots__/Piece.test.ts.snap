// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piece creates all combinations: brick 1`] = `
"brick:
-------------
1: rot:   0, flip: false
   |XX|
   |XX|
   |XX|

2: rot:  90, flip: false
   |XXX|
   |XXX|"
`;

exports[`Piece creates all combinations: bucket 1`] = `
"bucket:
-------------
1: rot:   0, flip: false
   |X X|
   |XXX|

2: rot:  90, flip: false
   |XX|
   |X |
   |XX|

3: rot: 180, flip: false
   |XXX|
   |X X|

4: rot: 270, flip: false
   |XX|
   | X|
   |XX|"
`;

exports[`Piece creates all combinations: cactus 1`] = `
"cactus:
-------------
1: rot:   0, flip: false
   |X |
   |XX|
   |X |
   |X |

2: rot:  90, flip: false
   |XXXX|
   |  X |

3: rot: 180, flip: false
   | X|
   | X|
   |XX|
   | X|

4: rot: 270, flip: false
   | X  |
   |XXXX|

5: rot:   0, flip: true
   | X|
   |XX|
   | X|
   | X|

6: rot:  90, flip: true
   |  X |
   |XXXX|

7: rot: 180, flip: true
   |X |
   |X |
   |XX|
   |X |

8: rot: 270, flip: true
   |XXXX|
   | X  |"
`;

exports[`Piece creates all combinations: elle 1`] = `
"elle:
-------------
1: rot:   0, flip: false
   |  X|
   |  X|
   |XXX|

2: rot:  90, flip: false
   |X  |
   |X  |
   |XXX|

3: rot: 180, flip: false
   |XXX|
   |X  |
   |X  |

4: rot: 270, flip: false
   |XXX|
   |  X|
   |  X|"
`;

exports[`Piece creates all combinations: fork 1`] = `
"fork:
-------------
1: rot:   0, flip: false
   |X |
   |X |
   |XX|
   | X|

2: rot:  90, flip: false
   | XXX|
   |XX  |

3: rot: 180, flip: false
   |X |
   |XX|
   | X|
   | X|

4: rot: 270, flip: false
   |  XX|
   |XXX |

5: rot:   0, flip: true
   | X|
   | X|
   |XX|
   |X |

6: rot:  90, flip: true
   |XX  |
   | XXX|

7: rot: 180, flip: true
   | X|
   |XX|
   |X |
   |X |

8: rot: 270, flip: true
   |XXX |
   |  XX|"
`;

exports[`Piece creates all combinations: hoe 1`] = `
"hoe:
-------------
1: rot:   0, flip: false
   |XX|
   |X |
   |X |
   |X |

2: rot:  90, flip: false
   |XXXX|
   |   X|

3: rot: 180, flip: false
   | X|
   | X|
   | X|
   |XX|

4: rot: 270, flip: false
   |X   |
   |XXXX|

5: rot:   0, flip: true
   |XX|
   | X|
   | X|
   | X|

6: rot:  90, flip: true
   |   X|
   |XXXX|

7: rot: 180, flip: true
   |X |
   |X |
   |X |
   |XX|

8: rot: 270, flip: true
   |XXXX|
   |X   |"
`;

exports[`Piece creates all combinations: snek 1`] = `
"snek:
-------------
1: rot:   0, flip: false
   |XX |
   | X |
   | XX|

2: rot:  90, flip: false
   |  X|
   |XXX|
   |X  |

3: rot:   0, flip: true
   | XX|
   | X |
   |XX |

4: rot:  90, flip: true
   |X  |
   |XXX|
   |  X|"
`;

exports[`Piece creates all combinations: utah 1`] = `
"utah:
-------------
1: rot:   0, flip: false
   |X |
   |XX|
   |XX|

2: rot:  90, flip: false
   |XXX|
   |XX |

3: rot: 180, flip: false
   |XX|
   |XX|
   | X|

4: rot: 270, flip: false
   | XX|
   |XXX|

5: rot:   0, flip: true
   | X|
   |XX|
   |XX|

6: rot:  90, flip: true
   |XX |
   |XXX|

7: rot: 180, flip: true
   |XX|
   |XX|
   |X |

8: rot: 270, flip: true
   |XXX|
   | XX|"
`;
